malta: add kernel 6.12 support

guidosarducci opened the pull request:
**Description**:

Add support for the `malta` target under kernel 6.12, and fix two package-related issues encountered during the work:
1. kselftests-bpf: update dependencies and clean up Makefile
2. zram: fix up menuconfig behaviour

**Testing**: This was regularly built and tested under QEMU for 4-5 months after rebasing on https://github.com/openwrt/openwrt/pull/16547 from @namiltd .

However, recent changes related to MIPS memory management (perhaps a9c0f28951fd1f8d6?) seem to cause problems.

On `malta/be32` there are many boot-time and run-time `WARNING`s (which disappear when the above commit is reverted):
```
[    1.314702] VFS: Mounted root (ext4 filesystem) readonly on device 8:0.
[    1.387469] Freeing unused kernel image (initmem) memory: 456K
[    1.387766] This architecture does not have kernel memory protection.
[    1.388570] Run /sbin/init as init process
[    1.622971] init: Console is alive
[    2.332498] kmodloader: loading kernel modules from /etc/modules-boot.d/*
[    2.385644] kmodloader: done loading kernel modules from /etc/modules-boot.d/*
[    2.404530] ------------[ cut here ]------------
[    2.405086] WARNING: CPU: 0 PID: 11 at mm/vmalloc.c:3361 vfree+0x1b8/0x330
[    2.406193] Trying to vfree() nonexistent vm area (83683000)
[    2.406831] Modules linked in: mii
[    2.408271] CPU: 0 UID: 0 PID: 11 Comm: kworker/0:1 Not tainted 6.12.25 #0
[    2.408754] Hardware name: mti,malta
[    2.409183] Workqueue: events do_free_init
[    2.409780] Stack : 80fd0000 83165cbc 00000000 801601b4 83165d2c 83165cfb 00000003 ffffffff
[    2.410100]         801dc21c 83045400 8300d600 6576656e 74730000 80a6ad98 83165cd8 83181500
[    2.410228]         00000000 00000000 80bb1364 83165b10 ffffefff 00000000 00000001 4109ae4c
[    2.410357]         83165d74 80b50000 811594b0 fffffffb 00000001 00000000 80bb1364 83165de8
[    2.410463]         00000000 803b8bc0 80fcbc88 82a65de0 00000000 00000000 00000000 81150000
[    2.410566]         ...
[    2.410807] Call Trace:
[    2.410848] [<8010918c>] show_stack+0x38/0x100
[    2.411419] [<80a6ad98>] dump_stack_lvl+0x70/0xb0
[    2.411474] [<801384dc>] __warn+0xb4/0x180
[    2.411515] [<80138738>] warn_slowpath_fmt+0x190/0x198
[    2.411552] [<803b8bc0>] vfree+0x1b8/0x330
[    2.411588] [<801dc284>] do_free_init+0x68/0x94
[    2.411624] [<8015b40c>] process_one_work+0x1e4/0x4d0
[    2.411668] [<8015d2b0>] worker_thread+0x328/0x498
[    2.411710] [<801676b0>] kthread+0x114/0x13c
[    2.411750] [<801031d8>] ret_from_kernel_thread+0x14/0x1c
[    2.412323]
[    2.412735] ---[ end trace 0000000000000000 ]---
[    2.429168] init: - preinit -
```

Also, I'm seeing a new build error on `malta/le64` which could be related:
```
arch/mips/kernel/module.c:136:13: error: 'free_phys' defined but not used [-Werror=unused-function]
  136 | static void free_phys(void *ptr)
      |             ^~~~~~~~~
cc1: all warnings being treated as errors
make[9]: *** [scripts/Makefile.build:229: arch/mips/kernel/module.o] Error 1
make[8]: *** [scripts/Makefile.build:478: arch/mips/kernel] Error 2
make[7]: *** [scripts/Makefile.build:478: arch/mips] Error 2
make[7]: *** Waiting for unfinished jobs....
```

Do these problems sound familiar to anyone working on 6.12? Thanks!

EDIT: Problems above are fixed by #18721.

**CC**: @hauke @Ansuel @PolynomialDivision

guidosarducci committed:
malta: 6.12: enable kernel for testing

All four subtargets build successfully and run under QEMU.

Signed-off-by: Tony Ambardar <itugrok@yahoo.com>
Link: https://github.com/openwrt/openwrt/pull/18712
Signed-off-by: Hauke Mehrtens <hauke@hauke-m.de>

guidosarducci committed:
kernel/malta: Create kernel files for v6.12 (from v6.6)

This is an automatically generated commit.

When doing `git bisect`, consider `git bisect --skip`.

Signed-off-by: Tony Ambardar <itugrok@yahoo.com>
Link: https://github.com/openwrt/openwrt/pull/18712
Signed-off-by: Hauke Mehrtens <hauke@hauke-m.de>

guidosarducci committed:
kernel/malta: Restore kernel files for v6.6

This is an automatically generated commit which aids following Kernel patch
history, as git will see the move and copy as a rename thus defeating the
purpose.

For the original discussion see:
https://lists.openwrt.org/pipermail/openwrt-devel/2023-October/041673.html

Signed-off-by: Tony Ambardar <itugrok@yahoo.com>
Link: https://github.com/openwrt/openwrt/pull/18712
Signed-off-by: Hauke Mehrtens <hauke@hauke-m.de>

guidosarducci committed:
kselftests-bpf: add support for kernel 6.12

Support new libpcap dependency and drop an old kernel version dependency.
Also improve formatting and avoid using unneeded deferred make variables.

Signed-off-by: Tony Ambardar <itugrok@yahoo.com>
Link: https://github.com/openwrt/openwrt/pull/18712
Signed-off-by: Hauke Mehrtens <hauke@hauke-m.de>

guidosarducci committed:
package/zram: fix up Kconfig behaviour for kernel 6.12

The initial 6.12 merge leaves package/zram menuconfig behaviour confusing.
Selecting "kmod-zram" opens a submenu of the available algorithms to build
and the actual algo to use. However, when "lzo-rle" is shown as the default
used, it still remains missing from the available list. Only enabling a
_different_ compression algo to build will also reveal "lzo-rle" as
available.

Update the Kconfig to show "lzo-rle" as available if used/referenced.

Fixes: 4708057e27 ("package/zram: update for kernel 6.12")
Signed-off-by: Tony Ambardar <itugrok@yahoo.com>
Link: https://github.com/openwrt/openwrt/pull/18712
Signed-off-by: Hauke Mehrtens <hauke@hauke-m.de>

DragonBluep commented:
You can try this patch https://github.com/DragonBluep/openwrt/commit/7eb9fff82e8854b30ec663fe0e458da6e8f5fed8.patch

guidosarducci commented:
> You can try this patch https://github.com/DragonBluep/openwrt/commit/7eb9fff82e8854b30ec663fe0e458da6e8f5fed8.patch

@DragonBluep Thanks! That works and resolves the problems with both `malta/le64` and `malta/be32`. How is your patch being merged? I can't seem to find a PR... Do you want me to include it here?

DragonBluep commented:
> > You can try this patch https://github.com/DragonBluep/openwrt/commit/7eb9fff82e8854b30ec663fe0e458da6e8f5fed8.patch
> 
> @DragonBluep Thanks! That works and resolves the problems with both `malta/le64` and `malta/be32`. How is your patch being merged? I can't seem to find a PR... Do you want me to include it here?

I’ve opened a new PR for it and added `Tested-by:` for you. https://github.com/openwrt/openwrt/pull/18721

guidosarducci commented:
Great! Tested-by is fine, so hope that gets merged soon. Thanks again.

namiltd commented:
Move [package/zram: fix up Kconfig behaviour for kernel 6.12](https://github.com/openwrt/openwrt/pull/18712/commits/ff9e275e1c72da1b2ec3c914971b73697a122cac) and [kselftests-bpf: add support for kernel 6.12](https://github.com/openwrt/openwrt/pull/18712/commits/53302df68f0daef0417a6480727e391628b94da8) to separate PRs.
It will be possible to have separate discussions.

guidosarducci commented:
@namiltd I don't believe that's necessary.

I actually raised the `package/zram` issue with you a few months ago in https://github.com/openwrt/openwrt/pull/16547#issuecomment-2608994894 but stopped seeing responses. Maybe some pictures would clarify what's confusing and needs fixing...

Currently, with `kmod-zram` selected, the method `lzo-rle` is listed as "Default" but _IS NOT_ shown as available or selected. Selecting one of the other available methods (e.g. `zstd`) will then make `lzo-rle` visible/available:
![Capture_1](https://github.com/user-attachments/assets/c77a25d4-e05f-4980-9890-a699b36277a1)
![Capture_2](https://github.com/user-attachments/assets/c5853022-dae0-408e-8036-30ed80660578)

After this patch. `lzo-rle` is correctly shown as available, and force-selected if used by default:
![Fix_Capture_1](https://github.com/user-attachments/assets/ce9ed669-cb22-4094-a42a-98772dab2bff)
![Fix_Capture_2](https://github.com/user-attachments/assets/1a9843f2-8933-4d15-949f-7af7a42f5707)

The `kselftests-bpf` package is one I created and maintain, and use to test malta updates. So no real reason to create unneeded PR churn and dependencies instead of a small number of related and properly documented commits in this PR.

@hauke Could you please have a look when you have a chance? This PR has been working for a while, and with 6.12 targets starting to trickle into the repo, it would be nice to have a baseline for MIPS.

namiltd commented on package/kernel/linux/modules/other.mk:
```diff
@@ -727,6 +724,12 @@ define KernelPackage/zram/config
+        config KERNEL_ZRAM_BACKEND_FORCE_LZO
+                def_bool !KERNEL_ZRAM_BACKEND_LZ4 && \
+                         !KERNEL_ZRAM_BACKEND_LZ4HC && \
+                         !KERNEL_ZRAM_BACKEND_ZSTD
```
Maybe add also !KERNEL_ZRAM_BACKEND_LZO because when KERNEL_ZRAM_BACKEND_LZO=y then KERNEL_ZRAM_BACKEND_FORCE_LZO is unnecessary.

namiltd replied:
Maybe add also !KERNEL_ZRAM_BACKEND_LZO because when KERNEL_ZRAM_BACKEND_LZO=y then KERNEL_ZRAM_BACKEND_FORCE_LZO is unnecessary (also remove select KERNEL_ZRAM_BACKEND_LZO).

namiltd commented:
> @namiltd I don't believe that's necessary.

It's not necessary, but most likely the 6.12 kernel support for Malta would have been approved already if it was separate.
So we wait for someone to analyze the ZRAM menu code.

guidosarducci committed:
malta: 6.12: update config after kernel refresh

Drop obsolete symbols GCC_ASM_GOTO_OUTPUT_WORKAROUND and HW_CONSOLE, and
PREEMPT_NONE_BUILD which is now set in top-level generic config.

Signed-off-by: Tony Ambardar <itugrok@yahoo.com>
Link: https://github.com/openwrt/openwrt/pull/18712
Signed-off-by: Hauke Mehrtens <hauke@hauke-m.de>
